### PR TITLE
Allow special character in MongoDB password

### LINKED ISSUE #13048: How to escape at-sign(@) in mongo config file
I have a MongoDB account that password includes at-sign(@)，the config file like this 
```
connector.name=mongodb_prod
mongodb.seeds=x.x.x.x:27017
mongodb.credentials=username:pass@word@admin
```
Then I found presto Mongo connector uses a SPLITTER to split string for creating a credential. So it will make connect fail if a password includes some special character.

How to escape the at-sign(@) in my presto config file?

https://github.com/prestodb/presto/blob/c223482fcc06a40c818feb06e9cf1670fb08c12a/presto-mongodb/src/main/java/com/facebook/presto/mongodb/MongoClientConfig.java#L134-L143

### PR DESCRIPTION
**Pull request by ebyhr**
Fix #13048

### CODE REVIEW

File: presto-mongodb/src/main/java/com/facebook/presto/mongodb/MongoClientConfig.java [hunk 1]
```diff
@@ -134,10 +133,18 @@ public MongoClientConfig setCredentials(String... credentials)
     private List<MongoCredential> buildCredentials(Iterable<String> userPasses)
     {
         ImmutableList.Builder<MongoCredential> builder = ImmutableList.builder();
-        for (String userPass : userPasses) {
-            List<String> values = USER_SPLITTER.splitToList(userPass);
-            checkArgument(values.size() == 3, "Invalid Credential format. Requires user:password@collection");
-            builder.add(createCredential(values.get(0), values.get(2), values.get(1).toCharArray()));
+        for (String userPassCollection : userPasses) {
```
Comment (mbasmanova) [inline presto-mongodb/src/main/java/com/facebook/presto/mongodb/MongoClientConfig.java]: - s/userPassCollection/userPass
- s/userPass/userAndPassword
- s/lastIndex/atPos
- s/firstIndex/colonPos

```
        for (String userPass : userPasses) {
            int atPos = userPass.lastIndexOf('@');
            checkArgument(atPos > 0, "Invalid Credential format. Requires user:password@collection");
            String userAndPassword = userPass.substring(0, atPos);
            String collection = userPass.substring(atPos + 1);

            int colonPos = userAndPassword.indexOf(':');
            checkArgument(colonPos > 0, "Invalid Credential format. Requires user:password@collection");
            String user = userAndPassword.substring(0, colonPos);
            String password = userAndPassword.substring(colonPos + 1);

            builder.add(createCredential(user, collection, password.toCharArray()));
        }
```

File: presto-mongodb/src/main/java/com/facebook/presto/mongodb/MongoClientConfig.java [hunk 2]
```diff
@@ -134,10 +133,18 @@ public MongoClientConfig setCredentials(String... credentials)
...
+            int lastIndex = userPassCollection.lastIndexOf('@');
+            checkArgument(lastIndex > 0, "Invalid Credential format. Requires user:password@collection");
```
Comment (mbasmanova) [inline presto-mongodb/src/main/java/com/facebook/presto/mongodb/MongoClientConfig.java]: - s/Credential/credential
- s/Requires/Required

File: presto-mongodb/src/main/java/com/facebook/presto/mongodb/MongoClientConfig.java [hunk 3]
```diff
@@ -134,10 +133,18 @@ public MongoClientConfig setCredentials(String... credentials)
...
+            String userPass = userPassCollection.substring(0, lastIndex);
+            String collection = userPassCollection.substring(lastIndex + 1);
+
+            int firstIndex = userPass.indexOf(':');
+            checkArgument(firstIndex > 0, "Invalid Credential format. Requires user:password@collection");
```
Comment (mbasmanova) [inline presto-mongodb/src/main/java/com/facebook/presto/mongodb/MongoClientConfig.java]: - s/Credential/credential
- s/Requires/Required

**Comment by mbasmanova** [comment]
I'm not seeing an existing test for `MongoClientConfig`, but there are tests for other configs that could be used as a reference, e.g. `TestCassandraClientConfig`.

**Comment by ebyhr** [comment]
@mbasmanova Thank you for your review. Updated.

File: presto-mongodb/src/main/java/com/facebook/presto/mongodb/MongoClientConfig.java [hunk 4]
```diff
@@ -135,9 +134,17 @@ public MongoClientConfig setCredentials(String... credentials)
     {
         ImmutableList.Builder<MongoCredential> builder = ImmutableList.builder();
         for (String userPass : userPasses) {
-            List<String> values = USER_SPLITTER.splitToList(userPass);
-            checkArgument(values.size() == 3, "Invalid Credential format. Requires user:password@collection");
-            builder.add(createCredential(values.get(0), values.get(2), values.get(1).toCharArray()));
+            int atPos = userPass.lastIndexOf('@');
```
Comment (highker) [inline presto-mongodb/src/main/java/com/facebook/presto/mongodb/MongoClientConfig.java]: nit: spell out words `atPosition` or `atIndex`

File: presto-mongodb/src/main/java/com/facebook/presto/mongodb/MongoClientConfig.java [hunk 5]
```diff
@@ -135,9 +134,17 @@ public MongoClientConfig setCredentials(String... credentials)
...
+            checkArgument(atPos > 0, "Invalid credential format. Required user:password@collection");
+            String userAndPassword = userPass.substring(0, atPos);
+            String collection = userPass.substring(atPos + 1);
+
+            int colonPos = userAndPassword.indexOf(':');
```
Comment (highker) [inline presto-mongodb/src/main/java/com/facebook/presto/mongodb/MongoClientConfig.java]: `colonPosition` or `colonIndex`

**Comment by mbasmanova** [comment]
@ebyhr Yuya, there are some remaining questions and comments here. Would you like to address them so we can merge this PR?

from @mbasmanova:

> The previous implementation had trimResults(). I assume the effect was that it would ignore leading and trailing spaces. The new implementation doesn't ignore these. Perhaps, add trim calls to keep the earlier behavior.

from @highker:

> nit: spell out words atPosition or atIndex 

> colonPosition or colonIndex

**Comment by ebyhr** [comment]
@mbasmanova Sorry for my late reply. When I tested about trim, I couldn't find the case breaking the backward compatibility. Could you tell me the test case? Or please let me know if I should add it even if there's no such cases.

**Comment by mbasmanova** [comment]
@ebyhr Yuya, the original code would process `user : password @ collection` and `user::password@@collection` strings the same way as `user:password@collection`. After the changes these strings will either raise an error or produce incorrect results. However, I don't know whether the original behavior is needed for real world scenarios. Hence, I would be Ok not matching that behavior.

**Pull request merged**